
app <- "numbat.rna.R"

args <- commandArgs(T)
if (length(args) < 6) {
  msg <- paste0("Usage: ", app, " <count matrix> <ref expression> \\
                                  <allele dataframe> <output dir> \\
                                  <output prefix> <ncores>")
  write(msg, file = stderr())
  quit("no", 1)
}

count_mtx_fn <- args[1]
ref_expr_fn <- args[2]
ale_fn <- args[3]
out_dir <- args[4]
out_prefix <- args[5]
ncores <- as.numeric(args[6])

library(numbat)

count_mtx <- readRDS(count_mtx_fn)
str(count_mtx) 

ref_expr <- readRDS(ref_expr_fn)
str(ref_expr)

ale <- readRDS(ale_fn)
str(ale)

obj <- run_numbat(
  count_mtx,    # gene x cell integer UMI count matrix 
  ref_expr,     # reference expression profile, a gene x cell type normalized expression level matrix
  ale,          # allele dataframe generated by pileup_and_phase script
  genome = "hg38",
  t = 1e-5,
  ncores = ncores,
  plot = TRUE,
  out_dir = out_dir,
  multi_allelic = FALSE     # https://github.com/kharchenkolab/numbat/issues/146
)

saveRDS(obj, sprintf("%s/%s.out.object.rds", out_dir, out_prefix))

print(paste0("[", app, "] All Done!"))

